### ISSUE TITLE
[MPIR-195] Attempting to configure maven-project-info-reports-plugin in reporting section of pom causes java.lang.ArrayIndexOutOfBoundsException to be thrown by site goal

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-project-info-reports-plugin#476]
**[Steve Gilbert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agency23)** opened **[MPIR-195](https://issues.apache.org/jira/browse/MPIR-195?redirect=false)** and commented

With an empty local repository using Maven 2.2.1, if plugin is added to the reporting section of the pom.xml, the site goal throws a java.lang.ArrayIndexOutOfBoundsException and the build fails.

Removing the plugin from the section or changing the version back to 2.1.2 (from 2.2) resolves the problem.

This prevents using any configuration of the plugin.

Placing the following pom into an empty directory and executing "mvn site" demonstrates the bug.

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>testgroup</groupId>
  <artifactId>testart</artifactId>
  <packaging>jar</packaging>
  <version>1.0-SNAPSHOT</version>

  <reporting>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-project-info-reports-plugin</artifactId>
        <version>2.2</version>
      </plugin>
    </plugins>
  </reporting>
</project>
```

---

**Affects:** 2.2

**Issue Links:**
- [MPIR-197](https://issues.apache.org/jira/browse/MPIR-197) ArrayIndexOutOfBoundsException on maven-2.2.1
 (_**"is duplicated by"**_)

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

which version of the site plugin are you using ?
2.1 is mandatory

**Comment by jira-importer** [comment]
**[Lukas Theussl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ltheussl)** commented

I put the above pom into an empty diretory, added site-plugin-2.1 in the build section and got no Exception.

**Comment by jira-importer** [comment]
**[Steve Gilbert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agency23)** commented

I am reopening this because I should not have to specify a version of the site plugin.  In the example pom I provided I specified version 2.2 for the project info reports plugin but if I do not specify a version, as I believe I should not have to, 2.2 is the version I get and the same error occurs.

The default version of the site plugin should work with the default version of the project info reports plugin.  The version of the site plugin maven is using is 2.0-beta-7 and the version of the project into reports plugin is version 2.2.

This is a valid bug.  Before the 2.2 version of this plugin was released a few days ago, I had a large number of poms that worked and now they do not.  How is this not a bug?

**Comment by jira-importer** [comment]
**[Lukas Theussl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ltheussl)** commented

There is no "default" version of the pir-plugin. It is not specified in any version of the maven super pom, so maven will use the last version it finds in your repo. OTOH, the site plugin version is specified in the super pom, so there is a "default" version, but it depends on which maven version you are running. In brief, your build is not reproducible (the fact that it works for you does not imply that it will also work for me), that's why you should always specify explicitly the versions of all plugins you are using [1]. This is all standard maven philosophy, if you have any questions about it please read up on the docs (admittedly often hard to find), or discuss it on the mailing list.

Finally, it has been specifically emphasized in the release notes that pir-2.2 requires site plugin 2.1. This is not a bug.

[1] http://maven.apache.org/guides/mini/guide-configuring-plugins.html

**Comment by jira-importer** [comment]
**[Steve Gilbert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agency23)** commented

Lukas- I am using version 2.2.1 of maven, which is the latest released, stable version AFAIK.

You say the site plugin's version is specified in the super pom, which establishes a "default" version.

You say 'There is no "default" version of the pir-plugin. It is not specified in any version of the maven super pom, so maven will use the last version it finds in your repo.'  I started with an empty repo.  Maven pulls down a version of the plugin into my empty repo.  Whatever it pulls down is the default version.  The fact that it is not declared in a pom does not mean there is no default version.  The default version is apparently determined by maven's resolution logic, but there is still a default version.

If you are going to say that I've not followed a maven best practice documented in an obscure document that that is your final position, I suppose there is nothing I can do but spend all day tomorrow modifying poms, however I still assert this is a regression bug.  I had a number of working poms that used the default versions of the plugins before the 2.2 version was released and now they don't work.

We've not declared our site plugin version so that we use the latest version released by your team.  I believe what you are saying is that in fact that practice is bad and that we should be tied to a specific version and relying on the latest released version is a bad idea.

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl@apache.org)** commented

Steve,

I think you are missing the point that Lukas is trying to make.

PIR Plugin 2.2 requires at least version 2.1 of the Site Plugin to work correctly. Unfortunately it is not possible for a plugin so say (in a programmatic way) that it requires a certain version of another plugin. What has been done instead is to document this requirement as much as possible, including mentioning it in the announcement.

An analogy would be that you were trying to use a new piece of software that requires Java 6, but you are trying it on Java 5. Do you consider that to be a bug in the software?

Your failure to follow the best practice that Lukas mentions, about always specifying your plugin versions, is the reason why you are having this problem now. Your builds will always be breakable if a new version comes out if you don't specify them. It usually takes an incident like this for people to start thinking some more about this.

**Comment by jira-importer** [comment]
**[Lukas Theussl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ltheussl)** commented

> Whatever it pulls down is the default version.

But this default version (as you define it) depends on your default repository and settings (if I happen to have an old version of pir in my local repo, maven will use that; if you start with an empty repo, maven will use the latest version it finds in whatever remote repos you have specified), and whenever we release a new version, maven will give you a new "default". So your build is not reproducible and you shouldn't be surprised if it breaks at one point.

> The fact that it is not declared in a pom does not mean there is no default version.

Yes it does. What is not in your pom is not defined.

> I still assert this is a regression bug.

No it isn't.

> We've not declared our site plugin version so that we use the latest version released by your team. I believe what you are saying is that in fact that practice is bad and that we should be tied to a specific version and relying on the latest released version is a bad idea.

Your experience proofs that it is indeed a very bad idea...

**Comment by jira-importer** [comment]
**[Steve Gilbert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agency23)** commented

I am not missing what you all are saying.  I understand completely but I disagree with the policy.

The analogy about an application that needs Java 6 to run doesn't hold for this situation.  Here is a more accurate analogy: I install a web browser and I expect that it will be able to render JPEG images on a web page, but it turns out the version of the graphics renderer the application uses is not specified and the one it chooses to use out of the box causes it to throw an exception and crash.  It's then on me to change the configuration so that a working version of the graphics render is used.  This makes no sense to me.  An application should perform its core task out of the box without requiring the user to configure it.

This case is the same.  Maven provides the site plugin out of the box through "mvn site".  The site plugin by default uses the project info reports plugin.  I don't have to configure either one of these plugins into maven.  They are there out of the box.  I don't think it's unreasonable to have an expectation that the application be configured out of the box to use the correct combination of these plugins.  If Maven did not compile java code out of the box without setting a plugin version first would that be a bug?

I realize you all disagree and believe that expectation is not reasonable.  I also realize you are not going to change your opinion about that.  I am going to change my poms because I have no other choice at this point.  I do appreciate your attention to and engagement on this issue.  I have learned that my assumption that Maven's core functionality will work out of the box without additional configuration was a poor one and I have adjusted that expectation.
